### ISSUE TITLE
ipynb+smart round trip not idempotent

### ISSUE THREAD

**Issue by ickc** [jgm/pandoc#7928]
MWE with pandoc 2.17.1.1 on macOS 12.2.1.

In `pandoc-bug.ipynb`,

```json
{
 "cells": [
 ],
 "metadata": {
  "title": "Number range 5–6"
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
```

Running `< pandoc-bug.ipynb | pandoc -s -f ipynb+smart -t ipynb+smart`,

```json
{
 "cells": [],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {
  "title": "Number range 5--6"
 }
}
```

Running it twice: `< pandoc-bug.ipynb | pandoc -s -f ipynb+smart -t ipynb+smart | pandoc -s -f ipynb+smart -t ipynb+smart`

```json
{
 "cells": [],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {
  "title": "Number range 5\\--6"
 }
}
```

And you can see each round trip will add a `\\`.

A similar construct in markdown doesn't result in this behavior, e.g. `echo '5–6' | pandoc -f markdown -t markdown | pandoc -f markdown -t markdown`.

**Comment by jgm** [comment]
Can you provide a complete pandoc-bug.ipynb (not just a fragment) so I can play with this?

**Comment by jgm** [comment]
A little easier to see the issue with:
```
% pandoc pandoc-bug.ipynb -s -f ipynb+smart -t native
Pandoc
  Meta
    { unMeta =
        fromList
          [ ( "jupyter"
            , MetaMap
                (fromList
                   [ ( "nbformat" , MetaString "4" )
                   , ( "nbformat_minor" , MetaString "5" )
                   , ( "title" , MetaString "Number range 5--6" )
                   ])
            )
          ]
    }
  []
```
We're getting `5--6` in the native instead of `5[UNICODE NDASH]6`.

**Comment by jgm** [comment]
Notes:

- the ipynb reader correctly parses this as `MetaString "Number range 5\8211\&6"` (whether or not smart is enabled)
- the ipynb writer outputs a unicode en-dash when smart is not enabled, but `--` when smart is enabled
- the markdown writer has the same behavior

Conclusion: the root problem is with the markdown writer.  The markdown writer should render `MetaString "Number range 5\8211\&6"` as a literal string with a unicode en-dash, rather than applying the "smart" transformation.

However, it's hard to see how we can fix this with current architecture.  The function that builds our yaml metadata block gets passed something like `[Str "5\8211\&6"]` and renders it using the markdown writer, which does the smart transformation.  We'd have to change the infrastructure around yaml metadata so that a separate function is used for MetaString.  Low priority for now.

**Comment by ickc** [comment]
> * the markdown writer has the same behavior

> Conclusion: the root problem is with the markdown writer.

Sorry don't quite follow here, `echo '5–6' | pandoc -f markdown -t markdown | pandoc -f markdown -t markdown` is ok?

(markdown has the `smart` extension enabled by default but ipynb doesn't.)

**Comment by jgm** [comment]
The problem arises because of the MetaString.  With `-f markdown -t markdown -s` we parse the metadata field as inlines.

**Comment by ickc** [comment]
I think it is the ipynb reader didn't convert this to en-dash?

Input: `pandoc-bug.ipynb`

```json
{
 "cells": [],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {
  "title": "Number range 5--6"
 }
}
```

```bash
❯ < pandoc-bug.ipynb | pandoc -s -f ipynb+smart -t native 
Pandoc
  Meta
    { unMeta =
        fromList
          [ ( "jupyter"
            , MetaMap
                (fromList
                   [ ( "nbformat" , MetaString "4" )
                   , ( "nbformat_minor" , MetaString "5" )
                   , ( "title" , MetaString "Number range 5--6" )
                   ])
            )
          ]
    }
  []
```

**Comment by jgm** [comment]
> I think it is the ipynb reader didn't convert this to en-dash?

Yes - as I was trying to explain, this is because it's in the context of a MetaString, so the content is treated as plain text and not parsed as markdown.

**Comment by ickc** [comment]
Right... Sorry, I didn't catch up earlier.

As usual with the ipynb ecosystem,
The spec at https://nbformat.readthedocs.io/en/latest/format_description.html didn't seem to define that,
but a "reference" implementation in nbconvert does:
https://nbconvert.readthedocs.io/en/latest/changelog.html?highlight=title#changes-in-how-we-handle-metadata
which isn't even well documented as it is only found in a release notes.

According to the source, it seems `title` and `authors` are used in LaTeX output:
https://github.com/jupyter/nbconvert/blob/5fe7c81f93b45f8f85ca01c15ac06cf7cb2d1a4e/share/jupyter/nbconvert/templates/latex/base.tex.j2
but only `title` in html output:
https://github.com/jupyter/nbconvert/blob/53c46d2785bfbaebdcb1ab7a959f7c09e9209477/share/jupyter/nbconvert/templates/classic/index.html.j2

**Comment by jgm** [comment]
This doesn't say that the title field is supposed to be parsed as markdown, does it?

**Comment by ickc** [comment]
I tested and the result is that markdown in title/authors are not interpreted in nbconvert. (i.e. nbconvert does not parse title or authors in metadata as markdown.)

**Comment by jgm** [comment]
Then that argues against the second solution above.  Unfortunately, the first one would be quite difficult given how things are currently arranged...

**Comment by jgm** [comment]
Actually, maybe this is simpler than I thought.

**Comment by jgm** [comment]
I'm going to call this closed by my last commit.

**Comment by ickc** [comment]
Thanks.

I used the latest nightly from
https://github.com/jgm/pandoc/actions/runs/1891833666
to test this but it seems the problem from the MWE still persists?

I used

```bash
$ < pandoc-bug.ipynb | pandoc-nightly-macos-2022-02-24/pandoc -s -f ipynb+smart -t ipynb+smart | pandoc -s -f ipynb+smart -t ipynb+smart
{
 "cells": [],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {
  "title": "Number range 5\\--6"
 }
}
```

**Comment by jgm** [comment]
The problem is that your second pandoc in the pipe is the system one, not the nightly.

**Comment by ickc** [comment]
Oh, right. How silly was me.
